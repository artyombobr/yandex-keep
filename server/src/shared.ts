export interface NotesProps {
  data: DataEntity;
}

export interface DataEntity {
  tags: TagsEntity[];
  colors: ColorsEntity[];
  notes: NotesEntity[];
}

export interface TagsEntity {
  id: number;
  tag: string;
}

export interface ColorsEntity {
  id: number;
  color: string;
}

export interface NotesEntity {
  type: string;
  title?: string;
  tags?: number[];
  color?: number;
  items?: ItemsEntity[];
  size?: string;
  created: number;
  text?: string;
  attachments?: AttachmentsEntity[];
  reminder?: number;
  url?: string;
  archive?: boolean;
}

export interface ItemsEntity {
  text: string;
  checked: boolean;
}

export interface AttachmentsEntity {
  type: string;
  url: string;
}
