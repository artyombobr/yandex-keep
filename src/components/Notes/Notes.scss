@import "../../shared";

.notes {
  margin: 0 16px;

  &__inner {
    display: grid;
    grid-template-columns: repeat(2, 1fr);
    grid-auto-flow: dense;
    grid-gap: 18px;
    margin: 0 auto
    //grid-template-columns: repeat(auto-fill, minmax(150px, 1fr));
  }
}

@media screen and (min-width: 768px) {
  .notes {
    &__inner {
      grid-template-columns: repeat(4, 1fr);
    }
  }
}

@media screen and (min-width: 1024px) {
  .notes {
    margin: 0 50px;

    &__inner {
      grid-template-columns: repeat(6, 1fr);
    }
  }
}

@media screen and (min-width: $large) {
  .notes {
    margin: 0 173px;
  }
}
