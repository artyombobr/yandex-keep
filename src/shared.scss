$small: 375px;
$medium: 769px;
$large: 1366px;

@function strip-unit($number) {
  @return $number / ($number * 0 + 1);
}

@function calc-size($f-min, $f-max, $v-min, $v-max) {
  $f-min: strip-unit($f-min);
  $f-max: strip-unit($f-max);
  $v-min: strip-unit($v-min);
  $v-max: strip-unit($v-max);
  $k: ($f-max - $f-min)/($v-max - $v-min);
  $b: $f-min - $k * $v-min;
  $b: $b * 1px;
  @return calc( #{$k} * 100vw + #{$b} );
}

@mixin fontsize($f-min, $f-max, $l-min, $l-max, $w-min, $w-max) {
  font-size: $f-min;
  @if $l-min !=0 {
    line-height: $l-min;
  }
  @media (min-width: $w-min+1) {
    font-size: calc-size($f-min, $f-max, $w-min, $w-max);
    @if $l-min !=0 {
      line-height: calc-size($l-min, $l-max, $w-min, $w-max);
    }
  }
  @media (min-width: $w-max) {
    font-size: $f-max;
    @if $l-max !=0 {
      line-height: $l-max;
    }
  }
}
