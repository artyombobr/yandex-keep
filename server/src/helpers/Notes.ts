import { NotesEntity } from '../shared';

class Notes {
  private readonly notes: NotesEntity[];

  private constructor(notes: NotesEntity[]) {
    this.notes = notes;
  }

  public static factory(notes: NotesEntity[]) {
    return new Notes(
      notes.map(note => {
        return {
          ...note,
          archive: false,
        };
      })
    );
  }

  public addNote(note: NotesEntity) {
    if (!note.size) {
      note = { ...note, size: Notes.getNoteSize(note) };
    }
    note = {
      ...note,
      id: this.notes.length,
      created: Date.now(),
      archive: false,
    };
    this.notes.unshift(note);
    return note;
  }

  public addToArchive(id: number): boolean {
    if (this.notes.length > id) {
      const index = this.notes.findIndex(note => note.id === Number(id));
      this.notes[index].archive = true;
      return true;
    }
    return false;
  }

  public editNote(id: number, note: NotesEntity): boolean {
    if (this.notes.length > id) {
      this.notes[id] = note;
      return true;
    }
    return false;
  }

  public deleteNote(id: number): boolean {
    if (this.notes.length > id) {
      this.notes.splice(id, 1);
      return true;
    }
    return false;
  }

  public toArray(): NotesEntity[] {
    return [...this.notes];
  }

  public map<T>(
    callback: (item: NotesEntity, index: number, array: NotesEntity[]) => T
  ): T[] {
    return this.notes.map(callback);
  }

  public filter(callback: (item: NotesEntity) => boolean) {
    return this.notes.filter(callback);
  }

  private static getNoteSize(note: NotesEntity) {
    if (
      (note.items && note.items.length > 20) ||
      (note.attachments && note.attachments.length > 3) ||
      (note.text && note.text.length > 400)
    ) {
      return 'l';
    }
    if (
      (note.items && note.items.length > 10) ||
      (note.attachments && note.attachments.length > 2) ||
      (note.text && note.text.length > 200)
    ) {
      return 'm';
    }
    return 's';
  }

  public *[Symbol.iterator]() {
    for (const note of this.notes) {
      yield note;
    }
  }
}

export default Notes;
