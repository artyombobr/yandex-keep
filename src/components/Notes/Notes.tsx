import React from 'react';
import { connect } from 'react-redux';
import {NotesEntity, NotesProps} from '../../shared';
import Note from '../Note/Note';
import './Notes.scss';

function Notes(props: any) {
  const {notes, colors} = props;
  return (
    <section className='notes'>
      <div className='notes__inner'>
        {notes.map((note: NotesEntity) => <Note key={note.created} note={note} colors={colors} />)}
      </div>
    </section>
  );
}

const mapStateToProps=(state: any)=>{
  return {
    notes: state.notes,
    colors: state.colors
  };
};

export default connect(mapStateToProps)(Notes);
