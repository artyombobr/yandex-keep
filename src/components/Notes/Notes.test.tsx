import React from 'react';
import { shallow } from 'enzyme';
import Notes from './Notes';
import { ColorsEntity } from '../../shared';

describe('Notes', () => {
  const notes = [
    {
      id: 1,
      type: 'text',
      text: 'Купить чайник с Bluetooth\nВидел у Xiaomi на днях',
      items: [
        {
          text: 'Оливочки',
          checked: false,
        },
      ],
      color: 0,
      tags: [0],
      size: 's',
      created: 1551593220000,
    },
  ];
  const colors: ColorsEntity[] = [
    {
      id: 0,
      color: '#E84747',
    },
  ];

  it('renders correctly', () => {
    const component = shallow(<Notes notes={notes} colors={colors} />);
    expect(component).toMatchSnapshot();
  });
});
