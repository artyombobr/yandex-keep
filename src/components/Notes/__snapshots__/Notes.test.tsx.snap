// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notes renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Notes
    colors={
      Array [
        Object {
          "color": "#E84747",
          "id": 0,
        },
      ]
    }
    notes={
      Array [
        Object {
          "color": 0,
          "created": 1551593220000,
          "id": 1,
          "items": Array [
            Object {
              "checked": false,
              "text": "Оливочки",
            },
          ],
          "size": "s",
          "tags": Array [
            0,
          ],
          "text": "Купить чайник с Bluetooth
Видел у Xiaomi на днях",
          "type": "text",
        },
      ]
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <div
        className="notes__inner"
      >
        <Note
          colors={
            Array [
              Object {
                "color": "#E84747",
                "id": 0,
              },
            ]
          }
          note={
            Object {
              "color": 0,
              "created": 1551593220000,
              "id": 1,
              "items": Array [
                Object {
                  "checked": false,
                  "text": "Оливочки",
                },
              ],
              "size": "s",
              "tags": Array [
                0,
              ],
              "text": "Купить чайник с Bluetooth
Видел у Xiaomi на днях",
              "type": "text",
            }
          }
        />
      </div>,
      "className": "notes",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Note
            colors={
              Array [
                Object {
                  "color": "#E84747",
                  "id": 0,
                },
              ]
            }
            note={
              Object {
                "color": 0,
                "created": 1551593220000,
                "id": 1,
                "items": Array [
                  Object {
                    "checked": false,
                    "text": "Оливочки",
                  },
                ],
                "size": "s",
                "tags": Array [
                  0,
                ],
                "text": "Купить чайник с Bluetooth
Видел у Xiaomi на днях",
                "type": "text",
              }
            }
          />,
        ],
        "className": "notes__inner",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "1551593220000",
          "nodeType": "function",
          "props": Object {
            "colors": Array [
              Object {
                "color": "#E84747",
                "id": 0,
              },
            ],
            "note": Object {
              "color": 0,
              "created": 1551593220000,
              "id": 1,
              "items": Array [
                Object {
                  "checked": false,
                  "text": "Оливочки",
                },
              ],
              "size": "s",
              "tags": Array [
                0,
              ],
              "text": "Купить чайник с Bluetooth
Видел у Xiaomi на днях",
              "type": "text",
            },
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
    "type": "section",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <div
          className="notes__inner"
        >
          <Note
            colors={
              Array [
                Object {
                  "color": "#E84747",
                  "id": 0,
                },
              ]
            }
            note={
              Object {
                "color": 0,
                "created": 1551593220000,
                "id": 1,
                "items": Array [
                  Object {
                    "checked": false,
                    "text": "Оливочки",
                  },
                ],
                "size": "s",
                "tags": Array [
                  0,
                ],
                "text": "Купить чайник с Bluetooth
Видел у Xiaomi на днях",
                "type": "text",
              }
            }
          />
        </div>,
        "className": "notes",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <Note
              colors={
                Array [
                  Object {
                    "color": "#E84747",
                    "id": 0,
                  },
                ]
              }
              note={
                Object {
                  "color": 0,
                  "created": 1551593220000,
                  "id": 1,
                  "items": Array [
                    Object {
                      "checked": false,
                      "text": "Оливочки",
                    },
                  ],
                  "size": "s",
                  "tags": Array [
                    0,
                  ],
                  "text": "Купить чайник с Bluetooth
Видел у Xiaomi на днях",
                  "type": "text",
                }
              }
            />,
          ],
          "className": "notes__inner",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "1551593220000",
            "nodeType": "function",
            "props": Object {
              "colors": Array [
                Object {
                  "color": "#E84747",
                  "id": 0,
                },
              ],
              "note": Object {
                "color": 0,
                "created": 1551593220000,
                "id": 1,
                "items": Array [
                  Object {
                    "checked": false,
                    "text": "Оливочки",
                  },
                ],
                "size": "s",
                "tags": Array [
                  0,
                ],
                "text": "Купить чайник с Bluetooth
Видел у Xiaomi на днях",
                "type": "text",
              },
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "div",
      },
      "type": "section",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
